src/Plugin/facets/query_type/HierarchyQueryType.php

```diff
@@ -0,0 +1,197 @@
+        // Set the options for the actual query.
+        $options = &$query->getOptions();
+        $options['search_api_facets'][$field_identifier] = $this->getFacetOptions();
+      }
+
+      // Add the filter to the query if there are active values.
+      $active_items = $this->facet->getActiveItems();
+
+      if (count($active_items)) {
+        if ($original_operator == HiearchicalMultiselectWidget::AND_WITH_HIERARCHY_OPERATOR) {
+          $operator = ListPresetFilter::AND_OPERATOR;
+        }
+        elseif ($original_operator == HiearchicalMultiselectWidget::OR_WITH_HIERARCHY_OPERATOR) {
+          $operator = ListPresetFilter::OR_OPERATOR;
+        }
+        elseif ($original_operator == HiearchicalMultiselectWidget::NONE_WITH_HIERARCHY_OPERATOR) {
+          $operator = ListPresetFilter::NOT_OPERATOR;
+        }
+        else {
+          $operator = $original_operator;
+        }
+
+        $filter = $query->createConditionGroup($operator, ['facet:' . $field_identifier]);
+        foreach ($active_items as $value) {
+          if (str_starts_with($value, '!(')) {
+            /** @var \Drupal\facets\UrlProcessor\UrlProcessorInterface $urlProcessor */
+            $urlProcessor = $this->facet->getProcessors()['url_processor_handler']->getProcessor();
+            foreach (explode($urlProcessor->getDelimiter(), substr($value, 2, -1)) as $missing_value) {
+              // Note that $exclude needs to be inverted for "missing".
+              // Maintain original logic.
+              if ($operator == $original_operator) {
+                $filter->addCondition($this->facet->getFieldIdentifier(), $missing_value, !$exclude ? '<>' : '=');
+              }
+              else {
+                $hierarchy_values = $this->getHierarchy($missing_value, $this->facet);
+                $filter->addCondition($this->facet->getFieldIdentifier(), $hierarchy_values, !$exclude ? 'NOT IN' : 'IN');
+              }
+
+            }
+          }
+          else {
+            if ($operator == $original_operator) {
+              $filter->addCondition($this->facet->getFieldIdentifier(), $value, $exclude ? '<>' : '=');
+            }
+            else {
+              $hierarchy_values = $this->getHierarchy($value, $this->facet);
+              $filter->addCondition($this->facet->getFieldIdentifier(), $hierarchy_values, $exclude ? 'NOT IN' : 'IN');
+            }
+          }
+        }
+        $query->addConditionGroup($filter);
+      }
+    }
+  }
+
+  /**
+   * Gets the hierarchy for the related field if support by its field type.
+   *
+   * @param string $value
+   *   The value to get the parents from.
+   * @param \Drupal\facets\FacetInterface $facet
+   *   The related facet.
+   *
+   * @return array
+   *   The hierarchy
+   */
+  protected function getHierarchy(string $value, FacetInterface $facet): array {
+    $facet_source = $facet->getFacetSource();
+    $field = $facet_source->getIndex()->getField($facet->getFieldIdentifier());
```

Comment:
If I learned everything in the last couple of days is that everywhere in code we check if the index actually has the field with that field identifier. So better to do a check to make sure $field is a field before continuing.